File: Problems/0230. (M) Kth Smallest Element in a BST/Python3/solution.py
class Solution:
  def kthSmallest(self, root: TreeNode, k: int) -> int:
    # inorder traversal with counter updates
    
    nodes = []
    cur = root
    # iterative traversal
    while True:
      if cur:
        nodes.append(cur)
        cur = cur.left
      elif nodes:
        cur = nodes.pop()
        # check and update counter appropriately
        if k == 1:
          return cur.val
        k -= 1
        cur = cur.right
      else:
        break

    # this line will never run given constraints on len(root) and k
    return -1

File: Problems/0230. (M) Kth Smallest Element in a BST/Java/solution.java
class Solution {
  public int kthSmallest(TreeNode root, int k) {
    // inorder traversal with counter updates

    Deque<TreeNode> nodes = new ArrayDeque();
    TreeNode cur = root;
    // iterative traversal
    while (true) {
      if (cur != null) {
        nodes.push(cur);
        cur = cur.left;
      }
      else if (nodes.size() != 0) {
        cur = nodes.pop();
        // check and update counter appropriately
        if (k == 1)
          return cur.val;
        k--;
        cur = cur.right;
      }
      else
        break;
    }
    // this line will never run given constraints on number of nodes and k
    return -1;
  }
}
